### PR TITLE
feat: add cosmwasm to umeed

### PR DESCRIPTION
**Pull request by RafilxTenfen**
## Description

- Add CosmWasm module to umeed

closes: #802 
----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by RafilxTenfen** [comment]
https://github.com/CosmWasm/wasmd/issues/105#issuecomment-612505124

**Comment by RafilxTenfen** [comment]
I will try to run some contracts [cw-plus](https://github.com/CosmWasm/cw-plus) to test this integration

**Comment by RafilxTenfen** [comment]
Talking with @brandoncomer, we probably are going to use the wasmd as permissioned as [Osmosis](https://github.com/osmosis-labs/osmosis/pull/769) did 

After testing around some contracts lime [cw20-basic](https://docs.cosmwasm.com/cw-plus/0.9.0/cw20/cw20-base-tutorial) I will change to be permissioned

**Comment by RafilxTenfen** [comment]
I was writing a script to test cw20 but got stuck running `sh tests/e2e/scripts/cosmwasm_contracts.sh` - `invalid address length; got: 32, max: 20`
image
@adamewozniak @toteki I am thinking that it has a way to convert the contract address... Any idea here?


> Accepting suggestions on this
>> My idea here was to create a new cw20, deploy (initiate), do some transfers, check the balances, approve this new token (CW20) as available for collateral, supply some amount and borrow umee against it

**Comment by toteki** [comment]
> My idea here was to create a new cw20, deploy (initiate), do some transfers, check the balances, approve this new token (CW20) as available for collateral, supply some amount and borrow umee against it

Note that for use as collateral, the oracle module (or mock oracle during testing) must have prices for both the borrowed and assets.

**Comment by RafilxTenfen** [comment]
> 

This error was because we have our own [function](https://github.com/umee-network/umee/blob/2da4a5b6a61a87ea7975622e72918a51b80bfc93/app/app.go#L983-L984) to verify address and it had check for size of 20, and the cosmwasm contract addr was 32 bytes

**Comment by RafilxTenfen** [comment]
Good thing, I had written some unit tests storing and instantiating cw20
After setting the coswasm as permissioned, broke the unit tests ;D

File: app/wasm_config.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	wasmtypes "github.com/CosmWasm/wasmd/x/wasm/types"
+	"github.com/cosmos/cosmos-sdk/codec"
+	servertypes "github.com/cosmos/cosmos-sdk/server/types"
+)
+
+const (
+	// DefaultUmeeWasmInstanceCost is initially set the same as in wasmd
+	DefaultUmeeWasmInstanceCost uint64 = 60_000
+	// DefaultUmeeWasmCompileCost set to a large number for testing
+	DefaultUmeeWasmCompileCost uint64 = 100
```
Comment (robert-zaremba) [inline app/wasm_config.go:17]: is 100 large?

Comment (RafilxTenfen) [reply app/wasm_config.go:17]: The wasm default is 3
```
	// DefaultCompileCost is how much SDK gas is charged *per byte* for compiling WASM code.
	// Benchmarks and numbers were discussed in: https://github.com/CosmWasm/wasmd/pull/634#issuecomment-938056803
	DefaultCompileCost uint64 = 3
```

**Comment by RafilxTenfen** [comment]
Tested cosmwasm upgrade at [umee-infra](https://github.com/umee-network/umee-infra/pull/116) noob scripts

**Comment by RafilxTenfen** [comment]
We decided to create initial custom hooks that query some information from our native modules
like.: `GetBorrow` and `GetExchangeRateBase`

**Comment by RafilxTenfen** [comment]
Custom hooks are working with these [smartcontracts examples](https://github.com/umee-network/umee-cosmwasm)

**Comment by RafilxTenfen** [comment]
Since CosmWasm is in the near-term Roadmap, I am going to open this PR to review as it is already pretty gigantic '-'

**Comment by adamewozniak** [comment]
Can we move this into another `app.go` file or use some sort of alternate build logic for the cosmwasm enabled version? Otherwise we're blocked on merging this until after mainnet

**Comment by RafilxTenfen** [comment]
This branch will stay as a draft until calypso is released

**Comment by RafilxTenfen** [comment]
> @RafilxTenfen can we close this? afaik cosmwasm won't be ready to integrate until 0.47

We can discuss that, one way to integrate (and still keep as a draft, but updated) would be to use one branch with 0.46 https://github.com/CosmWasm/wasmd/pull/938 
But don't know about this...

**Comment by RafilxTenfen** [comment]
@gsk967 Probably you are doing this in a new branch, take a look at `app/wasm/custom_plugins.go` 

I think we could close this PR but not delete the branch until the new umeed + cosmwasm works with https://github.com/umee-network/umee-cosmwasm

**Pull request closed without merging**